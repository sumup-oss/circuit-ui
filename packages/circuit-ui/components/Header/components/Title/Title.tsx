/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { css } from '@emotion/core';

import styled, { StyleProps } from '../../../../styles/styled';

export interface TitleProps {
  /**
   * The children component passed to the Title
   */
  children: React.ReactNode;
}

const baseStyles = ({ theme }: StyleProps) => css`
  label: header__title;
  font-size: ${theme.typography.headings.kilo.fontSize};
  line-height: ${theme.typography.headings.kilo.lineHeight};
  font-weight: ${theme.fontWeight.bold};
  color: ${theme.colors.n100};
  margin-left: ${theme.spacings.mega};
`;

const Title = styled('h1')<TitleProps>(baseStyles);

/**
 * @component
 */
export default Title;
