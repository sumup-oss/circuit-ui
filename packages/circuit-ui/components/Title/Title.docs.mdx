import { Status, Props, Story } from '../../../../.storybook/components';

# Title

<Status.Stable />

The Title component is used to render headings with large typography. Typically, large typography is intended for landing pages. In most cases, the [Headline](Typography/Headline) should be used instead.

<Story id="typography-title--base" />
<Props />

## Component variations

### Sizes

The Title component comes in four sizes. In most cases, use the [Headline component](Typography/Headline) component instead to render headings.

<Story id="typography-title--sizes" />

---

## Accessibility

All accessibility guidelines for the [Headline component](Typography/Headline) also apply to the Title component.
