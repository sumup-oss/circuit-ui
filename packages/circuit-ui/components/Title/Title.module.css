.base {
  font-weight: var(--cui-font-weight-bold);
  color: var(--cui-fg-normal);
  letter-spacing: -0.03em;
}

/* Sizes */

.one {
  font-size: var(--cui-typography-title-one-font-size);
  line-height: var(--cui-typography-title-one-line-height);
}

.two {
  font-size: var(--cui-typography-title-two-font-size);
  line-height: var(--cui-typography-title-two-line-height);
}

.three {
  font-size: var(--cui-typography-title-three-font-size);
  line-height: var(--cui-typography-title-three-line-height);
}

.four {
  font-size: var(--cui-typography-title-four-font-size);
  line-height: var(--cui-typography-title-four-line-height);
}
