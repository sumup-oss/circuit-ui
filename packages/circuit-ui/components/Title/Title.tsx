/**
 * Copyright 2021, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { css } from '@emotion/react';
import isPropValid from '@emotion/is-prop-valid';

import styled, { StyleProps } from '../../styles/styled';

type Size = 'one' | 'two' | 'three' | 'four';

export interface TitleProps {
  /**
   * A Circuit UI title size. Defaults to `one`.
   */
  size?: Size;
  /**
   * The HTML heading element to render.
   * Headings should be nested sequentially without skipping any levels.
   * Learn more at https://www.w3.org/WAI/tutorials/page-structure/headings/.
   */
  as: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
}

const baseStyles = ({ theme }: StyleProps) => css`
  font-weight: ${theme.fontWeight.bold};
  color: var(--cui-fg-normal);
  letter-spacing: -0.03em;
`;

const sizeStyles = ({ theme, size = 'one' }: StyleProps & TitleProps) => {
  if (!size) {
    return null;
  }

  return css`
    font-size: ${theme.typography.title[size].fontSize};
    line-height: ${theme.typography.title[size].lineHeight};
  `;
};

/**
 * A flexible title component capable of rendering any HTML heading element.
 */
export const Title = styled('h2', {
  shouldForwardProp: (prop) => isPropValid(prop) && prop !== 'size',
})<TitleProps>(baseStyles, sizeStyles);
