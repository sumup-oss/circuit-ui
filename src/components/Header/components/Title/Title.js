import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';

const baseStyles = ({ theme }) => css`
  label: header__title;
  font-size: ${theme.typography.headings.kilo.fontSize};
  line-height: ${theme.typography.headings.kilo.lineHeight};
  font-weight: ${theme.fontWeight.bold};
  color: ${theme.colors.n100};
  margin-left: ${theme.spacings.mega};
`;

const Title = styled('h1')(baseStyles);

Title.propTypes = {
  /**
   * The children component passed to the Title
   */
  children: PropTypes.node
};

/**
 * @component
 */
export default Title;
