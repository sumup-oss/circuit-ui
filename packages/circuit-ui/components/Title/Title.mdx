import { Meta, Status, Props, Story } from '../../../../.storybook/components';
import * as Stories from './Title.stories';

<Meta of={Stories} />

# Title

<Status.Stable />

The Title component is used to render headings with large typography. Typically, large typography is intended for landing pages. In most cases, the [Headline](Typography/Headline) should be used instead.

<Story of={Stories.Base} />
<Props />

## Component variations

### Sizes

The Title component comes in four sizes. In most cases, use the [Headline component](Typography/Headline) component instead to render headings.

<Story of={Stories.Sizes} />

---

## Accessibility

All accessibility guidelines for the [Headline component](Typography/Headline) also apply to the Title component.
