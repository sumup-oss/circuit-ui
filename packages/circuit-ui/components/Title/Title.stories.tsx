/**
 * Copyright 2021, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Title, TitleProps } from './Title';

export default {
  title: 'Typography/Title',
  component: Title,
};

export const Base = (args: TitleProps) => (
  <Title {...args}>This is a Title</Title>
);

const sizes = ['one', 'two', 'three', 'four'] as const;

export const Sizes = (args: TitleProps) =>
  sizes.map((s) => (
    <Title key={s} {...args} size={s}>
      This is a Title {s}
    </Title>
  ));
