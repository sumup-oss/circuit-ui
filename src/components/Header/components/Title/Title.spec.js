/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import Title from './Title';

describe('Title', () => {
  describe('styles', () => {
    it('should render with default styles', () => {
      const actual = create(<Title>Title</Title>);
      expect(actual).toMatchSnapshot();
    });
  });

  it('should render children', () => {
    const { getByTestId } = render(
      <Title>
        <span data-testid="child">Title</span>
      </Title>,
    );
    const childEl = getByTestId('child');
    expect(childEl).not.toBeNull();
    expect(childEl).toHaveTextContent('Title');
  });

  describe('accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(<Title>Title</Title>);
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
