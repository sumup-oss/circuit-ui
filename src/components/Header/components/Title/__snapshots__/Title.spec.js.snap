// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title styles should render with default styles 1`] = `
.circuit-0 {
  font-size: 17px;
  line-height: 24px;
  font-weight: 700;
  color: #F5F5F5;
  margin-left: 16px;
}

<h1
  class="circuit-0 circuit-1"
>
  Title
</h1>
`;
